nvme.c

```diff
@@ -288,47 +298,64 @@ static int get_telemetry_log(int argc, char **argv, struct command *cmd, struct
 		goto close_output;
 	}
 
-	err = write(output, (void *) hdr, 4096);
-	if (err != 4096) {
+	err = write(output, (void *) hdr, bs);
+	if (err != bs) {
 		fprintf(stderr, "Failed to flush all data to file!");
 		goto free_hdr;
 	}
-
-	num_blocks = max(hdr->dalb1, max(hdr->dalb2, hdr->dalb3));
-
-	full_size = num_blocks * 512;
-	/* Round to page boundary */
-	full_size += (4096 - (full_size % 4096));
-	/* We've already pulled 1 page worth of data, lets remove that. */
-	full_size -= 4096;
-	offset = 4096;
-
-	page_log = malloc(4096);
+	/* Setting offset at blocksize since the first block is the header
+	 * and is already pulled
+	 */
+	offset = bs;
+	page_log = malloc(bs);
 	if (!page_log) {
-		fprintf(stderr, "Failed to allocate %zu bytes for log\n",
-			full_size);
+		fprintf(stderr, "Failed to allocate %zu bytes for log\n", bs);
 		err = ENOMEM;
 		goto free_hdr;
 	}
 
-	while (full_size) {
-		err = nvme_get_telemetry_log(fd, page_log, 0, cfg.ctrl_init, 4096, offset);
+	switch (cfg.data_area) {
+	case 1:
+		full_size = (hdr->dalb1 * bs) + offset;
+		break;
+	case 2:
+		full_size = (hdr->dalb2 * bs) + offset;
+		break;
+	case 3:
+		full_size = (hdr->dalb3 * bs) + offset;
+		break;
+	default:
+		fprintf(stderr, "Invalid data area requested");
+		goto free_hdr;
+	}
+	/* In case full_size is somehow not a multiple of 512, return from this function
+	 * to prevent an infinite loop.
+	 */
+	if (full_size % bs) {
+		fprintf(stderr, "The full size of the data area requested is not rounded to the nearest block\n");
+		free(page_log);
```

Comment:
This one is OK, but it might be better to stick with policy of cleaning up memory at the end of function. 

renaming free_hdr into free_mem will do the trick, as well as moving free(page_log) one line below its current location.

this is just personal preference.

Reply:
Wouldn't putting free(page_log) inside of free_hdr and renaming it to free_mem give us issues if we attempt to goto it before page_log is malloced?

Reply:
as long as page_log is null, it's ok. 


https://linux.die.net/man/3/free
The free() function frees the memory space pointed to by ptr, which must have been returned by a previous call to malloc(), calloc() or realloc(). Otherwise, or if free(ptr) has already been called before, undefined behavior occurs. If ptr is NULL, no operation is performed.

Reply:
so you think initializing it to null would be better?

Reply:
yep, it'd nice, but it's up to you.

Reply:
I decided to just malloc page_log right after hdr, and then check for NULL on both right after. That way there's no need to initialize page_log to NULL, and there will never be a case where hdr is malloc'ed but page_log isn't.